### PR TITLE
Fix "Show error details" keeps showing after clicked

### LINKED ISSUE #19900: "Show error details" in Query Builder should be disabled after clicking
Steps to repro:

1. Run any query that fails.
2. See the `There was a problem with your question` error
3. Click `Show error details`
4. Error details are shown, but the `Show error details` link remains available active (i.e., it's still blue and hovering over it still adds the underline) and you can still click it.

I would expect that to get grayed out or give **some** sort of indication it has already been clicked after clicking it. It would be nice if it changed to a `Hide error details` button so you could hide them again, but I'm not super concerned about that.

I sat around for a minute or two trying to figure out why it wouldn't show me error details and kept clicking that button (thinking it was broken) before I realized that it was showing the error details (they were just not what I was expecting)

### PR DESCRIPTION
**Pull request by WiNloSt**
Fixes #19900 

## Repro steps
https://github.com/metabase/metabase/issues/19900#issuecomment-1057990108

## Before the fix

https://user-images.githubusercontent.com/1447303/156576467-443f1654-c398-4a58-84c6-05d1d3e2bad9.mp4

## After the fix

https://user-images.githubusercontent.com/1447303/156576481-79a0132f-5d26-4065-8ff0-001a638ddd47.mp4

### CODE REVIEW

File: frontend/src/metabase/components/ErrorDetails/ErrorDetails.tsx [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import React, { useState } from "react";
+import { t } from "ttag";
+import cx from "classnames";
+
+import { ErrorDetailsProps } from "./types";
+
+export default function ErrorDetails({
```
Comment (WiNloSt) [inline frontend/src/metabase/components/ErrorDetails/ErrorDetails.tsx:7]: Move the component to TypeScript + functional components + hooks

File: frontend/src/metabase/components/ErrorDetails/ErrorDetails.tsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+  function toggleShowError() {
+    setShowError(showError => !showError);
+  }
+
+  return (
+    <div className={className}>
+      <div className={centered ? "text-centered" : "text-left"}>
+        <a onClick={toggleShowError} className="link cursor-pointer">
+          {showError ? t`Hide error details` : t`Show error details`}
```
Comment (WiNloSt) [inline frontend/src/metabase/components/ErrorDetails/ErrorDetails.tsx:26]: Here's the changed logic.
- From setting `showError` to `true` 👉  Toggling `showError`
- From always showing `Show error details` 👉  Showing either `Hide error details` or `Show error details` based on the `showError` flag.

File: frontend/src/metabase/components/ErrorDetails/ErrorDetails.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import React, { useState } from "react";
```
Comment (WiNloSt) [inline frontend/src/metabase/components/ErrorDetails/ErrorDetails.tsx:1]: Removed `/* eslint-disable react/prop-types */` at the top of the file.

File: frontend/src/metabase/components/ErrorDetails/types.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+export interface ErrorDetailsProps {
+  details: Record<string, any>;
+  centered: boolean;
+  className: string;
```
Comment (daltojohnso) [inline frontend/src/metabase/components/ErrorDetails/types.ts:4]: Let's make `centered` and `className` optional.

And based on the usage of `details` in `ErrorDetails` around line 39, it looks like `details` is pretty loose. Maybe just make it `string | Record<string, any>` to account for the first predicate of the ternary (`typeof details === "string"`)

**Review by daltojohnso — CHANGES_REQUESTED** [review]

File: frontend/src/metabase/components/ErrorDetails/types.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+export interface ErrorDetailsProps {
```
Comment (WiNloSt) [inline frontend/src/metabase/components/ErrorDetails/types.ts:1]: @daltojohnso Fixed https://github.com/metabase/metabase/pull/20824#discussion_r818994361

**Review by daltojohnso — APPROVED** [review]

**Comment by WiNloSt** [comment]
@metabase-bot backport release-x.42.x

**Pull request merged**